Add handling for SIGTERM

pnuu opened the pull request:
This PR adds handling for SIGTERM signal, so that Trollflow2 can be terminated gracefully in for example container environment.

 - [ ] Closes #xxxx <!-- remove if there is no corresponding issue, which should only be the case for minor changes -->
 - [x] Tests added  <!-- for all bug fixes or enhancements -->
 - [x] Tests passed <!-- for all non-documentation changes -->
 - [x] Passes ``flake8 trollflow2`` <!-- remove if you did not edit any Python files -->
 - [x] Fully documented <!-- remove if this change should not be visible to users, e.g., if it is an internal clean-up, or if this is part of a larger project that will be documented later -->
 - [ ] Add your name to `AUTHORS.md` if not there already

pnuu commented:
There are currently two tests. One running `generate_messages` in a subprocess, another that does the same with `Runner`. With the latter case I get an error with `zmq.error.ZMQError: Invalid argument (addr='localhost:40000')` eventhough the `connection_parameters` are equal to the former. This causes the exit code to be `1` instead of the expected `0`.

Another problem is, that I'd expect the process to be up for 5 seconds due to the message iteration loop to wait for 5 seconds for new messages before timeout. Instead the process terminates immediately.

pnuu commented:
If I test with this script
```python
#!/usr/bin/env python

import os
import signal
import time


def main():
    """"""
    print(os.getpid())
    keep_looping = True
    tic = None

    def _signal_handler(signum, frame):
        nonlocal keep_looping
        nonlocal tic
        print("signal received")
        keep_looping = False
        tic = time.time()
    
    signal.signal(signal.SIGTERM, _signal_handler)

    while keep_looping:
        time.sleep(10)

    print("terminated with", signal.SIGTERM)
    print("time after termination", time.time() - tic)


if __name__ == "__main__":
    main()
```

and run `kill` from another terminal I see the termination time to be in relation to the `time.sleep()`. I thought these should be pretty equivalent.

pnuu commented:
Also using
```python
    while keep_looping:
        try:
            _ = in_queue.get(True, 10)
        except queue.Empty:
            pass
```
works as expected and the 10 second timeout is respected.

pnuu commented:
I'm guessing the premature termination is because of the `posttroll.listener.ListenerContainer` runs in a separate thread and is killed so the queue also exits.

I'll see if I can adapt the `test_sigterm_runner` so that there is a proper "workload" running and that it runs until completion.

pnuu commented:
With https://github.com/pytroll/trollflow2/pull/217/commits/11e5dff668a3ecf6d66fb12c466d4b94f3fe8156 the processing should be simulated and shown that we are waiting for it to finish before exiting.

I'll add a note about the SIGTERM signaling to documentation.

mraspaud commented on trollflow2/tests/test_launcher.py:
```diff
@@ -887,5 +887,69 @@ def test_generate_messages():
             assert msg.type in VALID_MESSAGE_TYPES
 
 
+def test_sigterm_generate_messages():
+    """Test that sending sigterm to Trollflow2 stops it."""
+    import os
+    import signal
+    from multiprocessing import Process
+
+    from trollflow2.launcher import generate_messages
+
+    connection_parameters = {"nameserver": False, "addresses": "localhost:40000", "topic": "/test"}
```
I've had problems to run tests with connections to localhost on some machines in the past. Would it be possible to switch this with an ipc connection instead?
eg:
```suggestion
def test_sigterm_generate_messages(tmp_path):
    """Test that sending sigterm to Trollflow2 stops it."""
    import os
    import signal
    from multiprocessing import Process

    from trollflow2.launcher import generate_messages
    ipc_path = os.fspath(tmp_path / "my_pipe")
    connection_parameters = {"nameserver": False, "addresses": f"ipc://{ipc_path}", "topic": "/test"}
```

pnuu replied:
The connection is never used, it's basically just saying "don't bother connecting to a nameserver that isn't there". I don't mind changing to your suggestion but I don't understand what it does.

mraspaud replied:
ipc is inter-process communication and basically just uses a pipe file on linux. Ie, doesn't need a network to work. It's supported by pyzmq and thus posttroll, I use it in other tests. It's treated as just another protocol.

mraspaud commented on trollflow2/tests/test_launcher.py:
```diff
@@ -887,5 +887,69 @@ def test_generate_messages():
+    time.sleep(5.0)
+
+
+@mock.patch("trollflow2.launcher.ListenerContainer")
```
Do we really need this mocking? would it work with posstroll.testing's `patched_subscriber_recv`? 
example usage: https://github.com/pytroll/pytroll-runner/blob/main/pytroll_runner/tests/test_runner.py#L323

pnuu replied:
How does it work replacing `ListenerContainer`, the underlying `Listener` and the output queue?

pnuu replied:
I tried this:
```python
    with patched_subscriber_recv([msg]):
        runner = Runner(product_list, connection_parameters)
    proc = Process(target=runner.run)
    proc.start()
    ...
```

but it exited immediately with exitcode 15 (SIGTERM) after the `kill` command.

mraspaud replied:
So `_fake_queue_logged_process`isn't called?

pnuu replied:
Yes, seems so, and the test completes in 1.007 seconds.

mraspaud replied:
Ok, thinking more about this, I would pass two messages to the patched subscriber and check that only one is processed.
But that doesn't help with the fake process being ignored. The patching is still there, right?
I don't understand why this is not working, the patching of the subscriber only passes messages down. Unless the ListenerContainer bypasses it somehow? Can you check if the message is consumed?

pnuu replied:
Ok, I think I figured it out. Things broke at https://github.com/pytroll/posttroll/blob/main/posttroll/listener.py#L47 because I had empty connection settings...

Switched to `patched_subscriber_recv` in https://github.com/pytroll/trollflow2/pull/217/commits/aefd2a33fd131f24065dcbafd123d968680c2406

But the `fake_queue_logged_process` is still not run. If I change it to sleep for 15 s the execution time doesn't change. But lets continue with that outside this thread.

mraspaud reviewed (commented):
Nice one, thanks for this. I have a couple of requests inline.
Then codecov is complaining about the signal_handler not being tested, but that can't be right, can it?

mraspaud commented on doc/source/launcher.rst:
```diff
@@ -11,3 +11,7 @@ messages by starting ``satpy_launcher.py`` with command-line arguments
 message publisher. Multiple publisher addresses can be given by supplying
 them with additional ``-a`` switches.
+
+The Trollflow2 process can be terminated gracefully by sending SIGTERM
+to the it, eg. by ``kill -SIGTERM <pid>`` where ``<pid>`` is the
+process ID number of Trollflow2.
```
One more thing: maybe we should explain that the gracefull termination implies finishing the current running processing?

pnuu replied:
Clarified in https://github.com/pytroll/trollflow2/pull/217/commits/5ecd049b70863022b727d343dea5ca77ae65e135

pnuu commented:
It is interesting that Codecov complains that `_signal_handler` is not covered, running all the tests locally with `pytest -s` prints out the signaling log message. I did try with `caplog` to catch those messages (the fixture seems to be still there in `test_sigterm_runner()`...), but I guess it didn't work because of the logs being in a separate thread/process and the messages are not propageted in my test case. I'll have another go.

pnuu commented:
Hmm, no, I can't capture the log messages even if I get them to print on the terminal. I also tried wrapping the `Process()` until `proc.join()` with `trollflow2.logging.logging_on` decorator but still only see the message on terminal and `caplog.text` stays empty.

pnuu commented:
And this is what is printed to terminal when using `with logging_on()` context:
```
[DEBUG: 2024-10-21 15:48:53 : trollflow2.launcher] Launching trollflow2 with subprocesses
[INFO: 2024-10-21 15:48:53 : trollflow2.launcher] New message received.
[DEBUG: 2024-10-21 15:48:53 : trollflow2.launcher] pytroll://my/topic file lahtinep@lahtinep21lk 2024-10-21T12:48:53.117334 v1.01 application/json {"filename": "foo"}
[INFO: 2024-10-21 15:48:54 : trollflow2.launcher] Caught signal to stop processing new messages and to terminate Trollflow2.
[INFO: 2024-10-21 15:48:59 : trollflow2.launcher] All 0 files produced nominally in 0:00:06.460422
```

mraspaud commented:
> And this is what is printed to terminal when using `with logging_on()` context:
> 
> ```
> [DEBUG: 2024-10-21 15:48:53 : trollflow2.launcher] Launching trollflow2 with subprocesses
> [INFO: 2024-10-21 15:48:53 : trollflow2.launcher] New message received.
> [DEBUG: 2024-10-21 15:48:53 : trollflow2.launcher] pytroll://my/topic file lahtinep@lahtinep21lk 2024-10-21T12:48:53.117334 v1.01 application/json {"filename": "foo"}
> [INFO: 2024-10-21 15:48:54 : trollflow2.launcher] Caught signal to stop processing new messages and to terminate Trollflow2.
> [INFO: 2024-10-21 15:48:59 : trollflow2.launcher] All 0 files produced nominally in 0:00:06.460422
> ```

yes, caplog and the queue logger sometimes ignore each other. Question is in which thread/process does the signal handling run? That maybe influences where the log goes?

pnuu commented:
> Question is in which thread/process does the signal handling run? That maybe influences where the log goes?

In `test_sigterm_generate_messages()` the signal handling is in the main thread, there's only the one process.

It gets more complicated in `test_sigterm_runner()`. `generate_messages()` is called in `launcher.Runner._get_message_iterator()` and the iterator is passed on to `_run_threaded()` or `_run_subprocess()`. The `queue_logged_process` is given a message from the message receiving iterator. The message receiving part should thus be within the main process of `Runner` that is started in the test case.

pnuu commented:
Looks like the messages do not end up in the `listener.output_queue`. If I revert https://github.com/pytroll/trollflow2/pull/217/commits/aefd2a33fd131f24065dcbafd123d968680c2406 the `_fake_queue_logged_process` is called again.

mraspaud commented:
> Looks like the messages do not end up in the `listener.output_queue`. If I revert [aefd2a3](https://github.com/pytroll/trollflow2/commit/aefd2a33fd131f24065dcbafd123d968680c2406) the `_fake_queue_logged_process` is called again.

ok, then I suppose the patched subscriber doesn't work with the listenercontainer. That is another issue that we need to fix in another PR, so feel free to put the mocking back.

pnuu commented:
Ok, I reverted from `patched_subscriber_recv` to mocking `ListenerContainer`. I also shortened the output queue wait time from 5 to 1 second to make testing a bit more streamlined. And removed unused `caplog` from `test_sigterm_runner()`.

mraspaud reviewed (approved):
LGTM
